src/core/Crust/Plugins/Plugin/hooks.ts

```diff
@@ -0,0 +1,454 @@
+  onRender?: (
+    options:
+      | {
+          width?: string | number;
+          height?: string | number;
+          extended?: boolean;
+        }
+      | undefined,
+  ) => void;
+  onResize?: (
+    width: string | number | undefined,
+    height: string | number | undefined,
+    extended: boolean | undefined,
+  ) => void;
+}): {
+  staticExposed: ((api: IFrameAPI) => GlobalThis) | undefined;
+  isMarshalable: Options["isMarshalable"] | undefined;
+  onPreInit: () => void;
+  onDispose: () => void;
+  onModalClose: () => void;
+  onPopupClose: () => void;
+} {
+  const ctx = usePluginContext();
+  const getLayer = useGet(layer);
+  const getBlock = useGet(block);
+  const getWidget = useGet(widget);
+
+  const event =
+    useRef<[Events<ReearthEventType>, EventEmitter<ReearthEventType>, (() => void) | undefined]>();
+
+  const pluginMessageSender = useCallback((msg: any) => {
+    event.current?.[1]("pluginmessage", msg);
+  }, []);
+
+  const onPreInit = useCallback(() => {
+    const e = events<ReearthEventType>();
+    let cancel: (() => void) | undefined;
+
+    if (ctx?.reearth.on && ctx.reearth.off && ctx.reearth.once) {
+      const source: Events<ReearthEventType> = {
+        on: ctx.reearth.on,
+        off: ctx.reearth.off,
+        once: ctx.reearth.once,
+      };
+      cancel = mergeEvents<ReearthEventType>(source, e[1], [
+        "cameramove",
+        "select",
+        "click",
+        "doubleclick",
+        "mousedown",
+        "mouseup",
+        "rightclick",
+        "rightdown",
+        "rightup",
+        "middleclick",
+        "middledown",
+        "middleup",
+        "mousemove",
+        "mouseenter",
+        "mouseleave",
+        "wheel",
+        "tick",
+        "resize",
+        "layeredit",
+      ]);
+    }
+
+    event.current = [e[0], e[1], cancel];
+
+    const instanceId = widget?.id ?? block?.id;
+    if (instanceId) {
+      ctx?.pluginInstances.addPluginMessageSender(instanceId, pluginMessageSender);
+    }
+  }, [
+    ctx?.reearth.on,
+    ctx?.reearth.off,
+    ctx?.reearth.once,
+    ctx?.pluginInstances,
+    widget?.id,
+    block?.id,
+    pluginMessageSender,
+  ]);
+
+  const onDispose = useCallback(() => {
+    event.current?.[1]("close");
+    event.current?.[2]?.();
+    event.current = undefined;
+    if (modalVisible) {
+      onPluginModalShow?.();
+    }
+    if (popupVisible) {
+      onPluginPopupShow?.();
+    }
+    const instanceId = widget?.id ?? block?.id;
+    if (instanceId) {
+      ctx?.pluginInstances.removePluginMessageSender(instanceId);
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [onPluginModalShow, onPluginPopupShow]);
+
+  const isMarshalable = useCallback(
+    (target: any) => defaultIsMarshalable(target) || !!mapRef?.current?.layers?.isLayer(target),
+    [mapRef],
+  );
+
+  const staticExposed = useMemo((): ((api: IFrameAPI) => GlobalThis) | undefined => {
```

Comment:
Only changed this object, and others were only copied from current layer system.